crates/noirc_evaluator/src/ssa/code_gen.rs

```diff
@@ -0,0 +1,1006 @@
+pub struct IRGenerator<'a> {
+    pub context: Option<&'a Context>,
+
+    pub first_block: arena::Index,
+    pub current_block: arena::Index,
+    pub blocks: arena::Arena<block::BasicBlock>,
+    pub nodes: arena::Arena<node::NodeObj>,
+    pub id0: arena::Index, //dummy index.. should we put a dummy object somewhere?
+    pub value_name: HashMap<arena::Index, u32>,
+    pub sealed_blocks: HashSet<arena::Index>,
+}
+
+impl<'a> IRGenerator<'a> {
+    pub fn new(context: &Context) -> IRGenerator {
+        let mut pc = IRGenerator {
+            context: Some(context),
+            id0: IRGenerator::dummy_id(),
+            first_block: IRGenerator::dummy_id(),
+            current_block: IRGenerator::dummy_id(),
+            blocks: arena::Arena::new(),
+            nodes: arena::Arena::new(),
+            // dummy_instruction: ParsingContext::dummy_id(),
+            value_name: HashMap::new(),
+            sealed_blocks: HashSet::new(),
+        }; //, objects: arena::Arena::new()
+        block::create_first_block(&mut pc);
+        pc
+    }
+
+    fn print_operand(&self, idx: arena::Index) -> String {
+        let var = self.get_object(idx);
+        if var.is_none() {
+            return format!("unknown {:?}", idx);
+        }
+        match var.unwrap() {
+            node::NodeObj::Obj(v) => v.print(),
+            node::NodeObj::Instr(i) => i.print_i(),
+            node::NodeObj::Const(c) => c.print(),
+        }
+    }
+
+    pub fn print_block(&self, b: &block::BasicBlock) {
+        for idx in &b.instructions {
+            let ins = self.get_instruction(*idx);
+            let mut str_res;
+            if ins.res_name.is_empty() {
+                str_res = format!("{:?}", idx);
+            } else {
+                str_res = ins.res_name.clone();
+            }
+            if ins.is_deleted {
+                str_res += " -DELETED";
+            }
+            let lhs_str = self.print_operand(ins.lhs);
+            let rhs_str = self.print_operand(ins.rhs);
+            let ins_str = format!(
+                "{} {} {}",
+                lhs_str,
+                format!(" op:{:?} ", ins.operator),
+                rhs_str
+            );
+            println!("{}: {}", str_res, ins_str);
+
+            if ins.operator == node::Operation::phi {
+                //    dbg!(&ins.phi_arguments);
+            }
+        }
+    }
+
+    pub fn print(&self) {
+        let mut i = 0;
+        for (_, b) in &self.blocks {
+            println!("************* Block n.{}", i);
+            i += 1;
+            self.print_block(b);
+        }
+    }
+
+    pub fn context(&self) -> &Context {
+        self.context.unwrap()
+    }
+
+    //Add an object to the nodes arena and set its id
+    pub fn add_object(&mut self, obj: node::NodeObj) -> arena::Index {
+        let idx = self.nodes.insert(obj);
+        let obj2 = self.nodes.get_mut(idx).unwrap(); //TODO-RIA can we avoid this? and simply modify obj?
```

Comment:
We can avoid it if we can query the next index of the nodes arena. Then we'd be able to delay the `self.nodes.insert(obj)` until the end of the function. I don't see such a function in the generational arena crate though. For now we may just have to deal with the extra lookup here and clones elsewhere.